File: PAT_Basic_Level_by_cpp/1039 到底买不买/1039 到底买不买.cpp
/*
 * 1039 到底买不买
 *
 * 【实现原理】
 * 店家有一串珠子（每个珠子有颜色），小红想要一串特定颜色序列的珠子。
 * 判断店家的珠子是否能满足小红的需求：
 * - 若能满足（即店家的珠子包含目标串中所有颜色的珠子），输出 Yes 及多余珠子数
 * - 若不能满足，输出 No 及缺少的珠子数
 *
 * 【算法思路】
 * 1. 统计店家各种颜色珠子的数量，存入计数数组 cnt
 * 2. 遍历目标串中的每个珠子：
 *    - 若 cnt 中对应颜色还有剩余，则消耗一个（cnt--）
 *    - 若 cnt 中对应颜色已无剩余，则 missing++
 * 3. 根据 missing 是否为 0 输出相应结果
 *
 * 【关键点】
 * - 使用大小为 256 的计数数组，覆盖所有 ASCII 字符
 * - 店家珠子总数 - 目标珠子总数 = 多余珠子数（当满足时）
 * - 珠子颜色由字母或数字表示，题目隐含字符范围在 ASCII 内
 *
 * 【复杂度分析】
 * - 时间复杂度：O(N + M)，N 为店家珠子数，M 为目标珠子数
 * - 空间复杂度：O(1)，固定 256 大小的数组
 */

#include <iostream>
#include <string>

using namespace std;

int main() {
    string shop, want;
    if (!(cin >> shop >> want)) return 0;

    int cnt[256] = {0};  // 统计店家每种颜色珠子的数量
    for (char c : shop) ++cnt[(unsigned char)c];

    int missing = 0;  // 缺少的珠子数
    for (char c : want) {
        if (cnt[(unsigned char)c] > 0) {
            --cnt[(unsigned char)c];  // 有该颜色珠子，消耗一颗
        } else {
            ++missing;  // 没有该颜色珠子，缺少数加一
        }
    }

    // missing == 0 表示所有珠子都满足
    if (missing == 0) {
        // 多余珠子数 = 店家的珠子总数 - 目标珠子总数
        cout << "Yes " << (shop.size() - want.size()) << endl;
    } else {
        cout << "No " << missing << endl;
    }
    return 0;
}

File: PAT_Basic_Level_by_java/1039 到底买不买/1039 到底买不买.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

/**
 * 1039 到底买不买
 *
 * 题目描述：
 * 给定店家现有的珠子串和顾客想要的珠子串，判断店家能否凑齐顾客所需的所有珠子。
 * 若能，输出"Yes"并返回多余珠子的数量；若不能，输出"No"并返回缺少珠子的数量。
 *
 * 实现原理：
 * 1. 使用 ASCII 码计数数组（大小 256）统计店家现有珠子串中每种字符的出现次数。
 * 2. 遍历想要的珠子串，对每个字符：
 *    - 若计数数组中该字符仍有剩余，则消耗一颗（计数减1）
 *    - 若计数数组中已无该字符，则 missing 计数加1
 * 3. 遍历结束后，若 missing == 0，说明可以凑齐，多余珠子数 = 店家串长度 - 想要串长度；
 *    否则输出缺少的珠子数。
 *
 * 时间复杂度：O(len(avail) + len(want))，仅需两次线性扫描
 */
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String avail = br.readLine();
        String want = br.readLine();
        if (avail == null) avail = "";
        if (want == null) want = "";
        // 使用 ASCII 计数数组统计店家有的珠子
        int[] cnt = new int[256];
        for (int i = 0; i < avail.length(); i++) {
            cnt[avail.charAt(i)]++;
        }
        // 遍历想要的珠子，检查是否够用
        int missing = 0;
        for (int i = 0; i < want.length(); i++) {
            char c = want.charAt(i);
            if (cnt[c] > 0) cnt[c]--;   // 有该珠子，消耗一颗
            else missing++;              // 缺少该珠子
        }
        // 根据是否缺少珠子输出结果
        if (missing == 0) {
            System.out.println("Yes " + (avail.length() - want.length()));
        } else {
            System.out.println("No " + missing);
        }
    }
}
